Sources/SwiftKafka/KafkaConsumer.swift

```diff
@@ -289,72 +271,285 @@ public final class KafkaConsumer {
+    ///
+    /// - Note: Invoking this function is not always needed as the ``KafkaConsumer``
+    /// will already shut down when consumption of the ``KafkaConsumerMessages`` has ended.
+    public func shutdownGracefully() {
+        let action = self.stateMachine.withLockedValue { $0.finish() }
+        switch action {
+        case .shutdownGracefullyAndFinishSource(let client, let source, let subscribedTopicsPointer, let logger):
+            self._shutdownGracefullyAndFinishSource(
+                client: client,
+                source: source,
+                subscribedTopicsPointer: subscribedTopicsPointer,
+                logger: logger
             )
+        case .none:
+            return
+        }
+    }
+
+    private func _shutdownGracefullyAndFinishSource(
+        client: KafkaClient,
+        source: Producer.Source,
+        subscribedTopicsPointer: UnsafeMutablePointer<rd_kafka_topic_partition_list_t>,
+        logger: Logger
+    ) {
+        source.finish()
+
+        let result = client.withKafkaHandlePointer { handle in
+            rd_kafka_consumer_close(handle)
         }
+
+        rd_kafka_topic_partition_list_destroy(subscribedTopicsPointer)
+
         guard result == RD_KAFKA_RESP_ERR_NO_ERROR else {
-            throw KafkaError.rdKafkaError(wrapping: result)
+            let error = KafkaError.rdKafkaError(wrapping: result)
+            logger.error("Closing KafkaConsumer failed: \(error.description)")
+            return
         }
-        return
     }
+}
 
-    /// Stop consuming messages. This step is irreversible.
-    func close() {
-        self.serialQueue.async {
-            guard !self.closed else {
-                return
+// MARK: - KafkaConsumer + StateMachine
+
+extension KafkaConsumer {
+    /// State machine representing the state of the ``KafkaConsumer``.
+    struct StateMachine {
+        /// The state of the ``StateMachine``.
+        enum State {
+            /// The state machine has been initialized with init() but is not yet Initialized
+            /// using `func initialize()` (required).
+            case uninitialized
+            /// We are in the process of initializing the ``KafkaConsumer``,
+            /// though ``subscribe()`` / ``assign()`` have not been invoked.
+            ///
+            /// - Parameter pollInterval: Amount of time between two subsequent polls invocations.
+            /// - Parameter client: Client used for handling the connection to the Kafka cluster.
+            /// - Parameter source: ``NIOAsyncSequenceProducer/Source`` used for yielding new elements.
+            /// - Parameter subscribedTopicsPointer: Pointer to a list of topics + partition pairs.
+            /// - Parameter logger: A logger.
+            case initializing(
+                pollInterval: Duration,
+                client: KafkaClient,
+                source: Producer.Source,
+                subscribedTopicsPointer: UnsafeMutablePointer<rd_kafka_topic_partition_list_t>,
```

Comment:
We should create an abstraction for this so we don't hold `UnsafeMutablePointer`s all the time.